config, net, test: asmap feature refinements and functional tests

jonatack opened the pull request:
This PR builds on PR #16702 to add functional tests / sanity checks and user-facing refinements for passing `-asmap` to configure ASN-based IP bucketing in addrman. As per our review discussion in that PR, the idea here is to handle aspects like functional tests and config arg handling that can help the PR be merged while enabling the author to focus on the bucketing itself.

- [x] add feature functional tests to verify node behaviour and debug log output when launching

  - `bitcoind` with no `-asmap` arg

  - `bitcoind -asmap=RELATIVE_FILENAME` to the unit test data skeleton asmap

  - `bitcoind -asmap` with no filename specified using the default asmap file

  - `bitcoind -asmap` with no filename specified and a missing default asmap file

- [x] add the ability to pass absolute path filenames to the `-asmap` config arg in addition to datadir-relative path filenames as per https://github.com/bitcoin/bitcoin/pull/16702#discussion_r361300447, and add test coverage

- [x] separate the asmap file finding and parsing checks, which allows adding tests for the case of a found but unparseable or empty asmap

- [x] add test for an empty asmap

- [x] various asmap fixups

- [x] move the asmap init code earlier in the init process to provide immediate feedback when passing an  `-asmap` config arg. This speeds up the `feature_asmap` functional test from 60 to 5 seconds! Credit to Wladimir J. van der Laan for the suggestion.

jonatack committed:
test: add feature_asmap functional tests

to verify node behaviour and debug log when launching bitcoind in these cases:

1. `bitcoind` with no -asmap arg, using /16 prefix for IP bucketing

2. `bitcoind -asmap=<relative path>`, using the unit test skeleton asmap

3. `bitcoind -asmap/-asmap=` with no file specified, using the default asmap

4. `bitcoind -asmap` with no file specified, and a missing default asmap file

The tests are order-independent. The slowest test (missing default asmap file)
is placed last.

jonatack commented:
Rebased to fix linter issues.

jonatack commented:
Rebased to fix linter and build-specific nits.

fanquake commented:
Shouldn't e4f5aa4b09a0f0c51def6b859609aee01b61d2df, b380e9a48e5c6cb7e1ed6558e7fbbde5feb0069b and eeb1a42341c9a5a9517a244506dd2a609962ee5c just be left as feedback/review on #16702? Especially if one of those commits is required for the CI to pass.

jonatack commented:
Yes, those are a temporary sketchpad for the moment, to be borrowed/used if useful. I have more important review questions on that PR that I'd like to concentrate on (the unit tests and actual bucketing) without adding further nit review at the moment.

The next commit will not be tests, which is why I tagged the PR with net as well.

jonatack committed:
config: use default value in -asmap config

and move to sorted position

jonatack committed:
config: enable passing -asmap an absolute file path

- allow passing an absolute file path to the -asmap config arg

- update the -asmap config help

- add a functional test in feature_asmap.py

jonatack committed:
config: separate the asmap finding and parsing checks

and update the tests.

jonatack committed:
logging: asmap logging and #include fixups

- move asmap #includes to sorted positions in addrman and init (move-only)

- remove redundant quotes in asmap InitError, update test

- remove full stops from asmap logging to be consistent with debug logging,
  update tests

jonatack committed:
test: add functional test for an empty, unparsable asmap

This is now testable after separating the asmap finding and parsing checks in
the previous commit.

jonatack commented:
The lone travis failure is unrelated and currently failing on all travis ci (macOS 10.14 import configparser).

laanwj commented:
+1 for allowing absolute `asmap` paths

jonatack committed:
init: move asmap code earlier in init process

and update feature_asmap.py and test_runner.py

This commit moves the asmap init.cpp code from the end of "Step 12: start node"
to "Step 6: network initialization" to provide feedback on passing an -asmap
config arg much more quickly. This change speeds up the feature_asmap.py
functional test file from 60 to 5 seconds by accelerating the 2 tests that use
`assert_start_raises_init_error`.

Credit to Wladimir J. van der Laan for the suggestion.

jonatack committed:
rpc: fix getpeerinfo RPCResult `mapped_as` type

and mention it is only available if the asmap config flag is set.

jonatack commented:
Rebased and added https://github.com/bitcoin/bitcoin/pull/16702#issuecomment-578736859 and https://github.com/bitcoin/bitcoin/pull/16702#discussion_r372364364 in the last 2 commits.

jonatack commented:
Rebased and added https://github.com/bitcoin/bitcoin/pull/16702#issuecomment-578736859 and https://github.com/bitcoin/bitcoin/pull/16702#discussion_r372364364 in the last 2 commits.

fanquake commented:
Can you squash a bunch of these? cd6d6d138f4844d35537f07c6c106e42dc0d8581 and b5cf2db1a4cd20175f37fc039c2bc0e15ccef1e4 could be combined. I don't think we need separate commits just to remove full stops and then add spaces to logging. 8554077e5532c26b28be95249992d96b68feebd7 reorders includes, which generally we don't do by itself, maybe combine into e44e8842b4da2829bbdcd982c83bb5ffb665b472.

gmaxwell commented:
Has anyone fuzzed the asmap file reader?  I anticipate users accepting asmaps from third parties...

gmaxwell commented:
Oh, nevermind, I just saw #18033

jonatack commented:
@fanquake thanks for having a look and the suggestions. Done, squashed all those down into e30a4691bacc232a984e049fb887b08860fa7c94.

jonatack commented:
Removed a link in the last commit that was polluting the parent PR comments.

laanwj commented:
Needs rebase (due to sipa's asmap changes in #18023, I think)

jonatack committed:
net: extract conditional to bool CNetAddr::IsHeNet

and remove redundant public declaration

jonatack commented:
Rebased, added 366aab9 for items found during review of other now-merged asmap PRs, and updated the new tests to use `self.chain` rather than 'regtest'.

jonatack commented on src/init.cpp:
```diff
@@ -409,7 +409,7 @@ void SetupServerArgs()
 
     gArgs.AddArg("-addnode=<ip>", "Add a node to connect to and attempt to keep the connection open (see the `addnode` RPC command help for more info). This option can be specified multiple times to add multiple nodes.", ArgsManager::ALLOW_ANY | ArgsManager::NETWORK_ONLY, OptionsCategory::CONNECTION);
-    gArgs.AddArg("-asmap=<file>", strprintf("Specify asn mapping used for bucketing of the peers (default: %s). Path should be relative to the -datadir path.", DEFAULT_ASMAP_FILENAME), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    gArgs.AddArg("-asmap=<file>", strprintf("Specify asn mapping used for bucketing of the peers (default: %s). Relative paths will be prefixed by a net-specific datadir location.", DEFAULT_ASMAP_FILENAME), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
```
Done. There are 3 other instances of "a net-specific datadir location" (`bitcoind -help | grep -B1 -A2 net-specific`) but I left them alone here.

jonatack commented on src/addrman.cpp:
```diff
@@ -6,16 +6,16 @@
     int tried_bucket = hash2 % ADDRMAN_TRIED_BUCKET_COUNT;
     uint32_t mapped_as = GetMappedAS(asmap);
-    LogPrint(BCLog::NET, "IP %s mapped to AS%i belongs to tried bucket %i.\n", ToStringIP(), mapped_as, tried_bucket);
+    LogPrint(BCLog::NET, "IP %s mapped to AS %i belongs to tried bucket %i\n", ToStringIP(), mapped_as, tried_bucket);
```
Thank you -- done.

jonatack commented on src/rpc/net.cpp:
```diff
@@ -83,7 +83,7 @@ static UniValue getpeerinfo(const JSONRPCRequest& request)
             "    \"addrbind\":\"ip:port\",    (string) Bind address of the connection to the peer\n"
             "    \"addrlocal\":\"ip:port\",   (string) Local address as reported by the peer\n"
-            "    \"mapped_as\":\"mapped_as\", (string) The AS in the BGP route to the peer used for diversifying peer selection\n"
+            "    \"mapped_as\":\"mapped_as\", (numeric) The AS in the BGP route to the peer used for diversifying peer selection\n"
```
Good catch -- done. Replaced with `\"mapped_as\": n,`

jonatack commented on src/addrman.cpp:
```diff
@@ -26,7 +26,7 @@ int CAddrInfo::GetNewBucket(const uint256& nKey, const CNetAddr& src, const std:
     int new_bucket = hash2 % ADDRMAN_NEW_BUCKET_COUNT;
     uint32_t mapped_as = GetMappedAS(asmap);
-    LogPrint(BCLog::NET, "IP %s mapped to AS%i belongs to new bucket %i.\n", ToStringIP(), mapped_as, new_bucket);
+    LogPrint(BCLog::NET, "IP %s mapped to AS %i belongs to new bucket %i\n", ToStringIP(), mapped_as, new_bucket);
```
Done.

jonatack commented:
Thank you @sipa for reviewing, updated as per `git diff 11a1028..1e26a1a`

```diff
diff --git a/src/addrman.cpp b/src/addrman.cpp
index ef34edc0a0..2f8a3a0bd5 100644
-    LogPrint(BCLog::NET, "IP %s mapped to AS %i belongs to tried bucket %i\n", ToStringIP(), mapped_as, tried_bucket);
+    LogPrint(BCLog::NET, "IP %s mapped to AS%i belongs to tried bucket %i\n", ToStringIP(), mapped_as, tried_bucket);
 
-    LogPrint(BCLog::NET, "IP %s mapped to AS %i belongs to new bucket %i\n", ToStringIP(), mapped_as, new_bucket);
+    LogPrint(BCLog::NET, "IP %s mapped to AS%i belongs to new bucket %i\n", ToStringIP(), mapped_as, new_bucket);
 
diff --git a/src/init.cpp b/src/init.cpp
index 6591852a7f..7b5e536b8f 100644
-    gArgs.AddArg("-asmap=<file>", strprintf("Specify asn mapping used for bucketing of the peers (default: %s). Relative paths will be prefixed by a net-specific datadir location.", DEFAULT_ASMAP_FILENAME), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    gArgs.AddArg("-asmap=<file>", strprintf("Specify asn mapping used for bucketing of the peers (default: %s). Relative paths will be prefixed by the net-specific datadir location.", DEFAULT_ASMAP_FILENAME), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);

diff --git a/src/rpc/net.cpp b/src/rpc/net.cpp
index 59d299e749..2f66fb9299 100644
-            "    \"mapped_as\":\"mapped_as\", (numeric) The AS in the BGP route to the peer used for diversifying peer selection\n"
+            "    \"mapped_as\": n,            (numeric) The AS in the BGP route to the peer used for diversifying peer selection\n"
+            "                                         (only available if the asmap config flag is set)\n"
```

pinheadmz commented on test/functional/feature_asmap.py:
```diff
@@ -0,0 +1,106 @@
+    def test_asmap_with_relative_path(self):
+        self.log.info('Test bitcoind -asmap=<relative path>')
+        self.stop_node(0)
+        name = 'ASN map'
```
Are you sure it's ok to have a space in a command line argument? Some OSes/shells might interpret this as `-asmap=ASN` and stop parsing, but I'm not sure

jonatack replied:
The CI builds didn't have an issue, but since there are no ACKs yet to lose, I can change it.

pinheadmz commented on test/functional/feature_asmap.py:
```diff
@@ -0,0 +1,106 @@
+        self.log.info('Test bitcoind -asmap with empty map file')
+        self.stop_node(0)
+        with open(self.default_asmap, "w", encoding="utf-8") as f:
+            f.write("")
```
I tried running this test with junk data here (`f.write("helloworld")`) bitcoind didn't mind interpreting that junk as a valid asmap file, didn't throw the expected parsing error. I suspect that's because the map file is just raw bytes, is there no way to check it for integrity?

jonatack replied:
I think that's outside the scope of this PR; perhaps as a follow-up.

pinheadmz commented:
ACK 1e26a1af96438edf96ecb2268569d5bec4a7f43e
optional nit https://github.com/bitcoin/bitcoin/pull/17812#discussion_r377729269 happy to reack if you decide to address that

jonatack commented:
Thanks @pinheadmz for reviewing, updated as per `git diff 1e26a1a..dccc6bf`:
```diff
test/functional/feature_asmap.py
-        name = 'ASN map'
+        name = 'ASN_map'
```

pinheadmz commented:
ACk dccc6bfc04e43bbdf8db35e0fb7288231b05973f

naumenkogs commented:
Great work, thanks! Especially those tests.
ACK dccc6bf

MarcoFalke commented:
ACK dccc6bfc04 🐔

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK dccc6bfc04 🐔
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUh6Cgv+M0g7bjlY1yqXorkNkyNeTx6YJpaqnM+w5VCirwJxVpH7hXGo8USv2t6d
N2u0oyAslzwupoD83ucutnx8yxbZ1cqDXvRasDX1I2pFACgKjSmd9NqkKOlpO6Qi
8KSSsvFeGBv2IO5dmIoRpDdUsebewBMnCdbJXh0g59+hkKo0GUU0NLcHIpdBgR1C
Un/ffCCmnF2KlvZAi1jje9fDc0ws2oMUzjb8NP7Zw2jtSEFev22j8lPpVX3efiWe
0l6+9eJ+La5B5OxwHx2afhEjybmlKSnEM5BidpYdDJRlN6/8GdkWACvCi9yr+uCV
NMfSVSV5lGXAGArf+lFH5j8ru1VelBy/GB2291q42b/wo/PvGMOn59IurgK4sfGr
/lzT66I+8Z8WtU4o8IDIShObkt4nJtHlosROun7Rx6XFC4enA5b82HkDnPfpY0vb
19r3DaJwTxQ22gEo1EhFAcKKTJThPbcwMtoNpsOMi+hzdBLunNTcCVbC2SN9ymoh
/PXwblvP
=zKAr
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `9e9a24d432005b3d1790df265d0b15fa6227860ab58337e3ed405bd9b7b94cba  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401089e9a24d432005b3d1790df265d0b15fa6227860ab58337e3ed405bd9b7b94cbaf01064ee4df0fdcb0f4a24440245bea4be1308fff0100f22d93eefbabf2894c897d3f833132808f1202ce3eab7f00b704bfe24ac1dd5329f20a41f313ebf19b64be2a3615ca7c7bc2308f1045e49f258f008eb8d89e11bf145ef0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0105f55bf600b31f8b9b3647dd04de66f5708f1045e49f259f008248c050600f0f0b10083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff01064543d66f1ad687bd0b08576b8c5847508f1045e49f259f008b13268571345c09e0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010ed0a4644d4f752b9823d523ec44265d008f02012909ecae05fc95ced8e2f4205bdab049dcde6195dba0eda7727ec0ed5cbf5e108f1045e49f259f008a970558b8854f7ce0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

jonatack commented:
Rebased. Thank you to the reviewers. Mind re-ACKing?

naumenkogs commented:
reACK d3ec323d151c5b19a7c4fa78000087266c9c6d2e

pinheadmz commented:
Confirming rebase didn't change anything:

```
$ git range-diff 712b7d9b4..dccc6bfc0  36f42e1bf..d3ec323d1
 1:  c4dc420cb =  1:  610cc93e9 test: add feature_asmap functional tests
 2:  919f4d7dd =  2:  a8a82d5bf config: use default value in -asmap config
 3:  429129440 =  3:  d7b1a5b04 config: enable passing -asmap an absolute file path
 4:  eacb2458e =  4:  b321be999 config: separate the asmap finding and parsing checks
 5:  0c3b1499b =  5:  366600e83 test: add functional test for an empty, unparsable asmap
 6:  89177df7a =  6:  d6e44f0b9 logging: asmap logging and #include fixups
 7:  392798830 =  7:  91ff62ff0 net: extract conditional to bool CNetAddr::IsHeNet
 8:  317bf6aea <  -:  --------- rpc: fix getpeerinfo RPCResult `mapped_as` type
 -:  --------- >  8:  ab8a6cbcb rpc: fix getpeerinfo RPCResult `mapped_as` type
 9:  dccc6bfc0 =  9:  d3ec323d1 init: move asmap code earlier in init process
```

```diff
<              "    \"addr\":\"host:port\",      (string) The IP address and port of the peer\n"
<              "    \"addrbind\":\"ip:port\",    (string) Bind address of the connection to the peer\n"
<              "    \"addrlocal\":\"ip:port\",   (string) Local address as reported by the peer\n"
< -            "    \"mapped_as\":\"mapped_as\", (string) The AS in the BGP route to the peer used for diversifying peer selection\n"
< +            "    \"mapped_as\": n,            (numeric) The AS in the BGP route to the peer used for diversifying peer selection\n"
< +            "                                         (only available if the asmap config flag is set)\n"
<              "    \"services\":\"xxxxxxxxxxxxxxxx\",   (string) The services offered\n"
<              "    \"servicesnames\":[              (array) the services offered, in human-readable form\n"
---
>              "    \"addr\" : \"host:port\",      (string) The IP address and port of the peer\n"
>              "    \"addrbind\" : \"ip:port\",    (string) Bind address of the connection to the peer\n"
>              "    \"addrlocal\" : \"ip:port\",   (string) Local address as reported by the peer\n"
> -            "    \"mapped_as\" : \"mapped_as\", (string) The AS in the BGP route to the peer used for diversifying peer selection\n"
> +            "    \"mapped_as\" : n,             (numeric) The AS in the BGP route to the peer used for diversifying peer selection\n"
> +            "                                           (only available if the asmap config flag is set)\n"
>              "    \"services\" : \"xxxxxxxxxxxxxxxx\",   (string) The services offered\n"
>              "    \"servicesnames\" : [              (json array) the services offered, in human-readable form\n"
```

reACK d3ec323d1

<details><summary>Sig:</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

reACK d3ec323d1
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAl5MCRYACgkQ5+KYS2KJ
yToEbA//eNfaa06Lt1kslQn4CgV6QBILrLFCcHct6/SMFDCGO1nHUAF+2m70j7Ro
PbgUaOKyuSZHreeVeoVfZtxJxjJnWhKUOugvEPCI0sMfezxznx9xSuYXxVCyJ1GG
e+sY9IWkTT/99blMqlSNAj1WVjr0p8LZlfBNDpbxJPd/zdTuqdFk0lMLsFA2/mdl
i9uDLsF9WxjfDdHb4rIL3CjUe22j+9idQuetkVZY/0BqwuQEjuSpVGeE/9ROkvyF
2RBUrIQfjcylq5NUv7NXBEX75xcvefsxv/fTKBbAU1yksKLckEAu87PLS5KscwCf
fCK5KydD/5f2bRE5Fi5t9VOmlTmj8WyMrvcd/Y3Wu/89hRQ0J70I8RYHP1xr+7TR
wgGL+Dwk4VPy34KEbZk9MZRvREm2xpqsKFl4/F8lUzYo4CYzYG/dhf2avpUUjS3O
198pydlsf6mIBXbWo2Qczzz5rzl334kY+g4T8I7B1IcZ/MlOd5RODOwCIBRXrFIN
xNJQ92S5lesLz+IkNFq0gMXp7ncPsSKEvdqyLeKD4BrKoSiS7T858fGgu/f1jnch
G+RZCXJf8fd2kF4bcu2fA/h8SiGD14NyULzPYvAog/bzqVwPBluSWr3dxRmpv6mr
sk31H1CbGGaFv31fNaVVv1zjZc1H32IpMty8TKhq0IQKqbac52c=
=s/w5
-----END PGP SIGNATURE-----
```

</details>

jonatack commented:
rebased

naumenkogs commented:
reACK 0b7ad4aba1c6a850d5ba927766ccd38cc003571c

pinheadmz commented:
Rebase looks cool

```
git range-diff 36f42e1bf..d3ec323d1 54a7ef612..0b7ad4aba
 1:  610cc93e9 =  1:  1fa86bb61 test: add feature_asmap functional tests
 2:  a8a82d5bf =  2:  499ff12f3 config: use default value in -asmap config
 3:  d7b1a5b04 =  3:  ca845a1aa config: enable passing -asmap an absolute file path
 4:  b321be999 =  4:  0ca73d89c config: separate the asmap finding and parsing checks
 5:  366600e83 =  5:  dc69b77e4 test: add functional test for an empty, unparsable asmap
 6:  d6e44f0b9 !  6:  74dea5777 logging: asmap logging and #include fixups
    @@ src/init.cpp
      #include <util/system.h>
      #include <util/threadnames.h>
      #include <util/translation.h>
    - #include <util/validation.h>
     -#include <util/asmap.h>
      #include <validation.h>
      #include <hash.h>
 7:  91ff62ff0 =  7:  295cd2b96 net: extract conditional to bool CNetAddr::IsHeNet
 8:  ab8a6cbcb =  8:  9ee5e8b0a rpc: fix getpeerinfo RPCResult `mapped_as` type
 9:  d3ec323d1 =  9:  0b7ad4aba init: move asmap code earlier in init process
```

reACK 0b7ad4aba1c6a850d5ba927766ccd38cc003571c

<details><summary>Sig</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

reACK 0b7ad4aba1c6a850d5ba927766ccd38cc003571c
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAl5dcNAACgkQ5+KYS2KJ
yTqiug//WdU8pZVxuxH2eF+Ky8+VUPynP7btq2cwqmhtPUcp/WMRRS+i8NAPcUWm
PP4DXaNaDbb5k/+/HmxWqju1v4kJRDxNVFzsK7KZ+515YKu6Vn3Zor/uV/XIKlwm
hOwNpC3w7ypoot2r8AbjZ8kWwPvUCdPqgXKt4p3VedxQ8XGsxnVmQ0/CpyI6Szb9
A4AmjoRW+VfEp8hOuhFTsE12BHfPJqGcskxLDiKVdyV31nE7U69yEj4T83lvBLUA
BoGysXQJKjuNbMYU9W7oGJsO/iX4Gu0hdRhYpbOjgCDM3p9AqSqDHDoYDNoHLA1c
Jk6qhwsK6HKde9Xd4tDrFxN6i2iTJ3d+Maw72Fnd2MKyapoWfQIefXU8pUnQN/A7
Kz6EpfTH+Oph089tQzJhRIkMapopV7gTVsGe1d9BL5/vMW9gQ5RjevOBPgY1pPwu
zxXNYdPeewF3Pb3qQ6Qq25RC8AHCE9OYuLl2F3i0UldDTRGY69AWNN7vxjYwCGPc
3EjGCKbzjas+iSaPyTwwcioZ+KN+N/ZVJC4f98RtcSCjeYOeESVWmt7mixen6QpH
3RrrCPPiJtmF9uB313xszgKEY+mDvv1iIeLj56eYvY4q4m/pJqCelOvS3X79E/OV
W1ngpN7Yq07x6zmlsa1H00bVBn68v7HIcIjwS49xxBPgHvLV2e4=
=9FZk
-----END PGP SIGNATURE-----

```


</details>

jonatack commented:
rebased

jonatack commented:
rebased due to #17809, second-to-last commit (5ba829e) is now
```diff
diff --git a/src/rpc/net.cpp b/src/rpc/net.cpp
index 482c689285..085921095e 100644
--- a/src/rpc/net.cpp
+++ b/src/rpc/net.cpp
@@ -86,7 +86,8 @@ static UniValue getpeerinfo(const JSONRPCRequest& request)
                             {RPCResult::Type::STR, "addr", "(host:port) The IP address and port of the peer"},
                             {RPCResult::Type::STR, "addrbind", "(ip:port) Bind address of the connection to the peer"},
                             {RPCResult::Type::STR, "addrlocal", "(ip:port) Local address as reported by the peer"},
-                            {RPCResult::Type::STR, "mapped_as", "The AS in the BGP route to the peer used for diversifying peer selection"},
+                            {RPCResult::Type::NUM, "mapped_as", "The AS in the BGP route to the peer used for diversifying\n"
+                                                                "peer selection (only available if the asmap config flag is set)"},
```

practicalswift commented:
In the sentence …

> "The AS in the BGP route to the peer …"

… isn't the correct technical term "AS path" rather than "BGP route" :)

jonatack commented:
Thanks @practicalswift, taking note for a follow-up (asmap release notes), at this point I prefer to preserve the 3 ACKs. Mind acking?

jonatack commented:
Can this PR be tagged for 0.20? 3 ACKs and feature freeze is in a week.

practicalswift commented:
@jonatack Sure! :)

ACK 1ba3e1cc21150abe632a5b82a1a38998b33815dc -- diff looks correct

fanquake reviewed (approved):
ACK 1ba3e1cc21150abe632a5b82a1a38998b33815dc
